packages/discovery-react-components/src/components/DocumentPreview/utils/useProvidedFile.ts

```diff
@@ -0,0 +1,64 @@
+import { useContext, useCallback, useState } from 'react';
+import { QueryResult } from 'ibm-watson/discovery/v2';
+import { SearchContext } from 'components/DiscoverySearch/DiscoverySearch';
+import useAsyncFunctionCall from 'utils/useAsyncFunctionCall';
+import { DocumentFile } from '../types';
+
+/**
+ * When the "file" prop is provided, just return it.
+ * Otherwise, fetch the file with the documentProvider.
+ * If the fetching time exceeds "fetchTimeout", nothing will be returned.
+ */
+export function useProvidedFile({
+  file,
+  document,
+  fetchTimeout
+}: {
+  file?: DocumentFile;
+  document?: QueryResult;
+  fetchTimeout?: number;
+}) {
+  const { documentProvider } = useContext(SearchContext);
+  const [fetchedFile, setFetchedFile] = useState<DocumentFile>();
+  const [fetching, setFetching] = useState(false);
+
+  useAsyncFunctionCall(
+    useCallback(
+      async (signal: AbortSignal) => {
+        if (!file && document && documentProvider) {
+          setFetching(true);
+          setFetchedFile(undefined);
+
+          const fetchedData = await Promise.race([
+            new Promise(async resolve => {
+              const hasFile = await documentProvider.provides(document!);
+              if (hasFile) {
+                const fetchedData = await documentProvider?.get(document!);
+                if (typeof fetchedData === 'string' || fetchedData instanceof ArrayBuffer) {
+                  resolve(fetchedData);
+                } else if (fetchedData?.type === 'pdf') {
+                  resolve(fetchedData.source);
```

Comment:
need `return` not to fall into the last `resolve(undefined)`
```suggestion
                  return resolve(fetchedData);
                } else if (fetchedData?.type === 'pdf') {
                  return resolve(fetchedData.source);
```

Reply:
Or, how about extracting the async function as a `fetchData` and put the result to the first element of the `Promise.race`?